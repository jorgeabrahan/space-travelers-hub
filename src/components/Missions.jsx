import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { missionsFetchAPI } from '../redux/missions/missionsSlice';
import Mission from './Mission';
import styles from '../css/Missions.module.css';

function Missions() {
  const dispatch = useDispatch();
  const { missions, status } = useSelector((store) => store.missions);

  useEffect(() => {
    if (missions.length > 0) return;
    if (status === 'failed') return;
    dispatch(missionsFetchAPI());
  }, [dispatch]);

  if (status !== 'succeeded') {
    return (
      <section className="center">loading...</section>
    );
  }
  return (
    <section className="center">
      <table className={styles.table}>
        <thead>
          <tr>
            <th className={styles.tableHeader}>Mission</th>
            <th className={styles.tableHeader}>Description</th>
            <th className={styles.tableHeader}>Status</th>
            <th className={styles.tableHeader}> </th>
          </tr>
        </thead>
        <tbody>
          {
            missions.missions.map((item, i) => (
              <Mission
                key={item.mission_id}
                missionName={item.mission_name}
                description={item.description}
                odd={i % 2 === 0}
              />
            ))
          }
        </tbody>
      </table>
    </section>
  );
}

export default Missions;
