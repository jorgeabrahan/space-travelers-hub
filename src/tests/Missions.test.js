import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Missions from '../components/Missions';

describe('Tests for <Missions /> component', () => {
  const name = 'Mission 1';
  const description = 'Mission description';
  const initialState = {
    missions: [{
      mission_id: '1',
      mission_name: name,
      description,
      reserved: false,
    }, {
      mission_id: '2',
      mission_name: name,
      description,
      reserved: true,
    }],
    status: 'succeeded',
    error: '',
  };

  const missionSlice = createSlice({
    name: 'missions',
    initialState,
  });

  const store = configureStore({
    reducer: {
      missions: missionSlice.reducer,
    },
  });

  const wrapper = (component) => (
    <Provider store={store}>
      {component}
    </Provider>
  );

  it('Should render two missions', () => {
    render(wrapper(<Missions />));
    expect(screen.queryByTestId('missions-container').children.length).toBe(2);
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(<Missions />)).toJSON()).toMatchSnapshot();
  });
});
