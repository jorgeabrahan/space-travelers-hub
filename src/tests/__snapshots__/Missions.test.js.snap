// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Missions /> component Should match the snapshot 1`] = `
<section
  className="center"
>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          className="tableHeader"
        >
          Mission
        </th>
        <th
          className="tableHeader"
        >
          Description
        </th>
        <th
          className="tableHeader"
        >
          Status
        </th>
        <th
          className="tableHeader"
        >
           
        </th>
      </tr>
    </thead>
    <tbody
      data-testid="missions-container"
    >
      <tr
        className="tableOdd"
      >
        <td
          className="tableMissionName"
        >
          Mission 1
        </td>
        <td
          className="tableMissionDescription"
        >
          Mission description
        </td>
        <td
          className="tableStatus"
        >
          <div
            className="notAMember"
          >
            not a member
          </div>
        </td>
        <td
          className="tableJoin"
        >
          <button
            className="outMission"
            onClick={[Function]}
            type="button"
          >
            Join Mission
          </button>
        </td>
      </tr>
      <tr
        className="tableEven"
      >
        <td
          className="tableMissionName"
        >
          Mission 1
        </td>
        <td
          className="tableMissionDescription"
        >
          Mission description
        </td>
        <td
          className="tableStatus"
        >
          <div
            className="member"
          >
            active member
          </div>
        </td>
        <td
          className="tableJoin"
        >
          <button
            className="onMission"
            onClick={[Function]}
            type="button"
          >
            Leave Mission
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</section>
`;
